### PR TITLE
Add user best practices documentation

### PR DESCRIPTION
**Pull request by dstrain115**
- Add document under 'start' about some basic best practices,
like using 'cirq.X' instead of 'cirq.ops.X'.
- Also added sections about immutability of gates, operations, and
  parameter resolvers.
- Removed name of former product manager.

Fixes:  #1373

### CODE REVIEW

**Comment by dstrain115** [comment]
FYI: test failure is a complex64 related flake

File: docs/start/best_practices.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+## What you see is what you get
+
+Cirq tries to be as true to the specified circuits as possible, especially with
+respect to hardware execution.  Cirq highly discourages any hidden automatic
+decomposition, compilation, or other modification of a circuit that is unknown
+to the user.  Any modification or transformation to the circuit should be
+initiated by the user of the library.
+
+This philosophy is important for many use cases.  For instance, certain
+benchmarking algorithms rely on the fact that gate sequences will not be transparently optimized, even if the circuit is nominally inefficient.
```
Comment (augustehirth) [inline docs/start/best_practices.md:61]: extremely minor nit: is "transparently" necessary? What does it mean in this context?

File: docs/start/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Of course, cirq provides routines and functions for compilation and
+transformation of circuits.  Users can and should call these routines.  However,
+cirq and resulting hardware integrations should not modify the circuits without
+the user's "permission".
+
+## Other style and performance guidelines
+
+*   Use `cirq.CircuitOperation` to more compactly define large, repeated
+    circuits.  This can save space and time for analysis of larger circuits.
+*   For hardware execution of multiple circuits, prefer using `run_sweep`  to
```
Comment (augustehirth) [inline docs/start/best_practices.md:72]: Link to /cirq/simulate/params here somewhere. `"Read more about how to do this in the [Parameter Sweeps](/cirq/simulate/params) page."`

**Review by augustehirth — COMMENTED** [review]
LGTM, and a great addition. Thanks for writing this.

File: docs/start/best_practices.md (hunk 1 above)
Comment (dstrain115) [reply docs/start/best_practices.md:61]: Removed.

File: docs/start/best_practices.md (hunk 2 above)
Comment (dstrain115) [reply docs/start/best_practices.md:72]: Done.

File: docs/start/best_practices.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+This page describes some of the best practices when using the Cirq library.
+Following these guidelines will help you write code that is more performant and
+less likely to break from version to version.  Many of these rules apply to
+other python libraries as well.
+
+## Use top-level constructs
+
+The Cirq library is designed so that important user-facing classes and objects
+are exposed at the package level.  Avoid referencing module names within cirq.
```
Comment (vtomole) [inline docs/start/best_practices.md:11]:
```suggestion
are exposed at the package level.  Avoid referencing module names within Cirq.
```

File: docs/start/best_practices.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+Cirq tries to be as true to the specified circuits as possible, especially with
+respect to hardware execution.  Cirq highly discourages any hidden automatic
+decomposition, compilation, or other modification of a circuit that is unknown
+to the user.  Any modification or transformation to the circuit should be
+initiated by the user of the library.
+
+This philosophy is important for many use cases.  For instance, certain
+benchmarking algorithms rely on the fact that gate sequences will not be optimized, even if the circuit is nominally inefficient.
+
+Of course, cirq provides routines and functions for compilation and
```
Comment (vtomole) [inline docs/start/best_practices.md:63]:
```suggestion
Of course, Cirq provides routines and functions for compilation and
```

File: docs/start/best_practices.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+transformation of circuits.  Users can and should call these routines.  However,
+cirq and resulting hardware integrations should not modify the circuits without
```
Comment (vtomole) [inline docs/start/best_practices.md:65]:
```suggestion
Cirq and resulting hardware integrations should not modify the circuits without
```

File: docs/start/best_practices.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+While python's flexibility allows developers to modify just about anything, it
+is bad practice to modify classes that are designed to be immutable.  Doing so
+can violate assumptions made in other parts of the library.
+
+In particular, attributes of `cirq.Gate`, `cirq.Operation`, `cirq.Moment`, and
+`cirq.ParamResolver` should not be modified after creation.  If these objects
+need to be modified, a new object should be created instead.
+
+Violating this priniciple could cause problems in other parts of the code.  For
```
Comment (vtomole) [inline docs/start/best_practices.md:34]:
```suggestion
Violating this principle could cause problems in other parts of the code.  For
```

**Review by vtomole — COMMENTED** [review]
Some nits

**Comment by dstrain115** [comment]
@vtomole Hmmm... looks like there is some problem with your CLA that is now blocking the PR since I took your suggestions.

**Comment by pavoljuhas** [comment]
LGTM.  There is also a [best_practices.ipynb](https://github.com/quantumlib/Cirq/blob/master/docs/google/best_practices.ipynb) notebook,
but I suppose it is clear enough which best practices are which.

**Review by pavoljuhas — APPROVED** [review]

**Comment by vtomole** [comment]
> Hmmm... looks like there is some problem with your CLA that is now blocking the PR since I took your suggestions.

Oh that's very new. Lemme look if i changed my default email.

**Review by vtomole — APPROVED** [review]
I don't know if CLAs expire, but I just resigned it. LGTM

**Pull request merged**